qiskit/providers/experiment/experiment_data.py

```diff
@@ -0,0 +1,1006 @@
+        self._metadata = copy.deepcopy(metadata)
+        self._source = self._metadata.pop(
+            "_source",
+            {
+                "class": f"{self.__class__.__module__}.{self.__class__.__name__}",
+                "metadata_version": self._metadata_version,
+                "qiskit_version": qiskit_version(),
+            },
+        )
+
+        self._service = None
+        self._backend = backend
+        self.auto_save = False
+        self._set_service_from_backend(backend)
+
+        self._id = experiment_id or str(uuid.uuid4())
+        self._type = experiment_type
+        self._tags = tags or []
+        self._share_level = share_level
+        self._notes = notes or ""
+
+        self._jobs = ThreadSafeOrderedDict(job_ids or [])
+        self._job_futures = ThreadSafeList()
+        self._errors = []
+
+        self._data = ThreadSafeList()
+        self._figures = ThreadSafeOrderedDict(figure_names or [])
+        self._analysis_results = ThreadSafeOrderedDict()
+
+        self._deleted_figures = deque()
+        self._deleted_analysis_results = deque()
+
+        self._created_in_db = False
+        self._extra_data = kwargs
+
+    def _set_service_from_backend(self, backend: Union[Backend, BaseBackend]) -> None:
+        """Set the service to be used from the input backend.
+
+        Args:
+            backend: Backend whose provider may offer experiment service.
+        """
+        with contextlib.suppress(Exception):
+            self._service = backend.provider().service("experiment")
+        with contextlib.suppress(Exception):
+            self.auto_save = self._service.option("auto_save")
+
+    def add_data(
+        self,
+        data: Union[Result, List[Result], Job, List[Job], Dict, List[Dict]],
+        post_processing_callback: Optional[Callable] = None,
+        **kwargs: Any,
+    ):
+        """Add experiment data.
+
+        Note:
+            This method is not thread safe and should not be called by the
+            `post_processing` function.
+
+        Args:
+            data: Experiment data to add.
+                Several types are accepted for convenience:
+
+                    * Result: Add data from this ``Result`` object.
+                    * List[Result]: Add data from the ``Result`` objects.
+                    * Job: Add data from the job result.
+                    * List[Job]: Add data from the job results.
+                    * Dict: Add this data.
+                    * List[Dict]: Add this list of data.
+            post_processing_callback: Callback function invoked when all pending
+                jobs finish. This ``ExperimentData`` object is the only argument
+                to be passed to the callback function.
+            **kwargs: Keyword arguments to be passed to the callback function.
+        Raises:
+            TypeError: If the input data type is invalid.
+        """
+        if isinstance(data, dict):
+            self._add_single_data(data)
```

Comment:
@jyu00 What about the post-processing callback function, in the cases where the data is not a `Job` ?

Reply:
fixed in fc53978